pkg/controller/servicebinding/binder.go

```diff
@@ -152,55 +166,68 @@ func (b *binder) removeSpecVolumes(
 		return nil, err
 	}
 	volumes = b.removeVolumes(volumes)
-	if err = unstructured.SetNestedSlice(obj.Object, volumes, b.getVolumesPath()...); err != nil {
+
+	volumesUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(volumes)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = unstructured.SetNestedMap(obj.Object, volumesUnstructured, b.getVolumesPath()...); err != nil {
 		return nil, err
 	}
 	return obj, nil
 }
 
 // updateVolumes inspect informed list assuming as []corev1.Volume, and if binding volume is already
 // defined just return the same list, otherwise, appending the binding volume.
-func (b *binder) updateVolumes(volumes []interface{}) ([]interface{}, error) {
+func (b *binder) updateVolumes(volumes []corev1.Volume) ([]interface{}, error) {
 	name := b.sbr.GetName()
 	log := b.logger
+	var indices []int
+	var updatedVolumes []interface{}
 
 	log.Debug("Checking if binding volume is already defined...")
-	for _, v := range volumes {
-		volume, ok := v.(map[string]interface{})
-		if !ok {
-			return nil, err.New("type asserting volume to map of string to interface{} error")
-		}
-		if name == volume["name"] {
+	for k, v := range volumes {
+		if v.Name == name {
 			log.Debug("Volume is already defined!")
-			return volumes, nil
+			if v.Secret.SecretName == b.sbr.Status.Secret {
+				u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(v)
+				if err != nil {
+					return nil, err
+				}
+				return append(updatedVolumes, u), nil
+			}
+			indices = append(indices, k)
 		}
 	}
 
-	log.Debug("Appending new volume.", "Secret.Name", name)
+	for _, v := range indices {
+		volumes = append(volumes[:v], volumes[v+1:]...)
```

Comment:
Can't this be removing the wrong items after the first modification into `volumes`? Assuming `indices` is ordered, the second index would be off by one.

I'd create another `[]corev1.Volume` variable and copy all the indices not in `indices` to not change `volume` to avoid mistakes and add unit tests to cover the this scenario.